File: 115. Distinct Subsequences/115. Distinct Subsequences.rb
# @param {String} s
# @param {String} t
# @return {Integer}
def num_distinct(s, t)
    m, n = s.length, t.length
    dp = Array.new(m + 1) { Array.new(n + 1, 0) }
    
    # Empty string is a subsequence of empty string once
    (0..m).each { |i| dp[i][0] = 1 }
    
    # Fill dp table
    (1..m).each do |i|
        (1..n).each do |j|
            dp[i][j] = dp[i-1][j]
            dp[i][j] += dp[i-1][j-1] if s[i-1] == t[j-1]
        end
    end
    
    dp[m][n]
end

File: 115. Distinct Subsequences/115. Distinct Subsequences.c
int numDistinct(char* s, char* t) {
    int m = strlen(s), n = strlen(t);
    unsigned long long dp[1001][1001] = {0};
    
    // Empty string is a subsequence of empty string once
    for (int i = 0; i <= m; i++)
        dp[i][0] = 1;
    
    // Fill dp table
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            dp[i][j] = dp[i-1][j];
            if (s[i-1] == t[j-1])
                dp[i][j] += dp[i-1][j-1];
        }
    }
    
    return dp[m][n];
}